src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,845 @@
+            new_hidden_state[0, :, : hidden_state_0.shape[2]] = hidden_state_0
+            new_hidden_state[1, :, : hidden_state_1.shape[2]] = hidden_state_1
+            hidden_states = hidden_states + (new_hidden_state,)
+        else:
+            hidden_states = hidden_states + (torch.cat([hidden_state_0, hidden_state_1]),)
+    return hidden_states
+
+
+def stack_hidden_states_list(hidden_states: List[torch.Tensor]) -> torch.Tensor:
+    """
+    Given a list of hidden states tensors, stack them together using torch.stack.
+    We assume that the hidden states are of shape (batch_size, hidden_state_size, num_keypoints).
+    If all hidden states have the same shape, we stack them along the batch dimension:
+    [hidden_state_0, hidden_state_1, ...] -> torch.stack([hidden_state_0, hidden_state_1, ...], dim=0)
+    If the hidden states have different shapes, the smaller ones will be padded with zeros:
+    (batch_size * 2, hidden_state_size, max(num_keypoints_0, num_keypoints_1))
+    """
+    all_hidden_state_shape_the_same = True
+    first_attention_prob_shape = hidden_states[0].shape
+    for attention_prob in hidden_states[1:]:
+        if attention_prob.shape != first_attention_prob_shape:
+            all_hidden_state_shape_the_same = False
+    if all_hidden_state_shape_the_same:
+        stacked_hidden_state = torch.stack(hidden_states, dim=0)
+    else:
+        stacked_hidden_state = torch.zeros(
+            len(hidden_states),
+            2,
+            hidden_states[0].shape[1],
+            max([hidden_state.shape[2] for hidden_state in hidden_states]),
+        )
+        for i, hidden_state in enumerate(hidden_states):
+            stacked_hidden_state[i, :, :, : hidden_state.shape[2]] = hidden_state
+    return stacked_hidden_state
+
+
+def batch_hidden_states(
+    hidden_states: Union[List[torch.Tensor], List[Tuple[torch.Tensor]]]
+) -> Union[List[torch.Tensor], List[Tuple[torch.Tensor]]]:
+    """
+    Given a list of hidden states, batch them together using torch.stack.
+    We assume that the hidden states are of shape (batch_size, hidden_state_size, num_keypoints).
+    The list can be of two forms:
+    - List of hidden states: [hidden_state_0, hidden_state_1, ...]
+    - List of tuples of hidden states: [(hidden_state_0_0, hidden_state_1_0, ...), (hidden_state_0_1, hidden_state_1_1, ...), ...]
+    If the list is of the first form, we stack the attention probabilities along the batch dimension:
+    [hidden_state_0, hidden_state_1, ...] -> torch.stack([hidden_state_0, hidden_state_1, ...], dim=0)
+    If the list is of the second form, we stack the attention probabilities along the batch dimension for each tuple:
+    [(hidden_state_0_0, hidden_state_1_0, ...), (hidden_state_0_1, hidden_state_1_1, ...), ...] ->
+    [torch.stack([hidden_state_0_0, hidden_state_1_0, ...], dim=0), torch.stack([hidden_state_0_1, hidden_state_1_1, ...], dim=0), ...]
+    """
+    if len(hidden_states) > 0:
```

Comment:
Do we need to handle all of these cases?

Reply:
Same comment as above, fixed in [this commit](https://github.com/huggingface/transformers/pull/29886/commits/25fd37d15e16a2600a7c0eaa052890d81177580a)